application/comit_node/src/bam_api/mod.rs

```diff
@@ -1,134 +1,141 @@
 use crate::{
-    bam_api::header::{Error, FromBamHeader, Header, ToBamHeader},
-    swap_protocols::SwapProtocols,
+    bam_ext::{FromBamHeader, ToBamHeader},
+    comit_client::SwapDeclineReason,
+    swap_protocols::{
+        asset::AssetKind,
+        ledger::{Bitcoin, Ethereum, LedgerKind},
+        SwapProtocols,
+    },
 };
+use bam::json::Header;
+use ethereum_support::Erc20Token;
 
-pub mod header;
 pub mod rfc003;
 
-mod ledger_impls {
-    use crate::{
-        bam_api::header::{Error, FromBamHeader, Header, ToBamHeader},
-        swap_protocols::ledger::{Bitcoin, Ethereum},
-    };
-
-    impl FromBamHeader for Bitcoin {
-        fn from_bam_header(mut header: Header) -> Result<Self, Error> {
-            header.has_value("Bitcoin")?;
-
-            Ok(Bitcoin::new(header.parameter("network")?))
-        }
-    }
-
-    impl ToBamHeader for Bitcoin {
-        fn to_bam_header(&self) -> Result<Header, Error> {
-            Ok(Header::with_value("Bitcoin").with_parameter("network", self.network)?)
-        }
-    }
-
-    impl FromBamHeader for Ethereum {
-        fn from_bam_header(mut header: Header) -> Result<Self, Error> {
-            header.has_value("Ethereum")?;
-
-            Ok(Ethereum::new(header.parameter("network")?))
-        }
-    }
-
-    impl ToBamHeader for Ethereum {
-        fn to_bam_header(&self) -> Result<Header, Error> {
-            Ok(Header::with_value("Ethereum").with_parameter("network", self.network)?)
-        }
+impl FromBamHeader for LedgerKind {
```

Comment:
❓Might be a stupid question, but why is this not a `impl From<..> for ...` ?  Is it because we need to track the errors?

Reply:
also, since `Header` is a `bam::json::Header` (BamHeader), maybe you can skip the `Bam` in the name of the trait?

Reply:
> questionMight be a stupid question, but why is this not a `impl From<..> for ...` ? Is it because we need to track the errors?

Yes.

Reply:
> also, since `Header` is a `bam::json::Header` (BamHeader), maybe you can skip the `Bam` in the name of the trait?

Same argument would work for `Header` :D 
I thought being explicit makes more sense.

Reply:
@bonomat there are no stupid questions!